network/application/MQTT_eclipse_paho/mqtt_noos_support.c

```diff
@@ -0,0 +1,187 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include "mqtt_noos_support.h"
+#include <stdlib.h>
+#include "timer.h"
+#include "error.h"
+#include "util.h"
+#include "delay.h"
+
+/******************************************************************************/
+/**************************** Global Variables ********************************/
+/******************************************************************************/
+
+/* Timer reference used by the functions */
+static struct timer_desc	*timer;
+
+/* Number of timer references */
+static uint32_t			nb_references;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Initialize porting for mqtt. Must be called from \ref mqtt_init
+ * @param extra_init_param - Platform specific param to initialize a timer_desc
+ * @return
+ *  - \ref SUCCESS : On success
+ *  - \ref FAILURE : Otherwise
+ */
+int32_t mqtt_noos_init(void *extra_init_param)
+{
+	struct timer_init_param init_param = {
+		.id = 0,
+		.freq_hz = 1000,
+		.load_value = 0,
+		.extra = extra_init_param
+	};
+	int32_t ret;
+
+	ret = timer_init(&timer, &init_param);
+	if (IS_ERR_VALUE(ret)) {
+		timer = NULL;
+		return FAILURE;
+	}
+
+	ret = timer_start(timer);
+	if (IS_ERR_VALUE(ret)) {
+		timer_remove(timer);
+		timer = NULL;
+		return FAILURE;
+	}
+
+	nb_references++;
+
+	return SUCCESS;
+}
+
+
+/**
+ * @brief Remove resources allocated with \ref mqtt_noos_init.
+ *
+ * Must be called from \ref mqtt_remove
+ */
+void mqtt_noos_remove()
+{
+	nb_references--;
+	if (!nb_references) {
+		timer_remove(timer);
+		timer = NULL;
+	}
+}
+
+/* Implementation of TimerInit used by MQTTClient.c */
+void TimerInit(Timer* t)
+{
+	UNUSED_PARAM(t);
```

Comment:
Maybe move the noos timer init here?

Reply:
This function is called from the paho MQTT and there is no remove for this in order to delete the resources allocated if we use timer_init_here. That's way I used a global timer and the functions mqtt_noos_init and mqtt_noos_remove to handle it.